new/missing-method-missing.md

```diff
@@ -0,0 +1,156 @@
+---
+RFC: unassigned
+Author: John Keiser <john@johnkeiser.com>
+Status: Draft
+Type: Standards Track
+---
+
+# Remove magic and method_missing from Chef::Resource lookup
+
+This proposal:
+
+- Switches to explicit methods for Chef recipe DSL instead of method_missing
+- Removes all special treatment of classes in the Chef::Resource and Chef::Provider namespaces
+- Automatically adds resource DSL for all named descendants of Chef::Resource, no matter what their namespace
+
+## Motivation
+
+    As a Chef user,
+    I want resources declared in my own namespace to be in recipe DSL automatically,
+    so that I don't have to write the name of the resource twice.
+
+    As a Chef user,
+    I want to declare resources in my own namespace,
+    So that I get better error messages when there is a resource conflict.
+
+    As a Chef developer,
+    I want meaningful names in stack traces for methods,
+    So that I can actually read the debug output.
+
+    As a Chef developer,
+    I want less if statements, case statements and metaprogramming in DSL method calls,
+    So that I can debug and follow logic more easily.
+
+## Specification
+
+### Create explicit DSL methods for resources
+
+When looking up resources, there is no method anywhere that corresponds to the
+actual resource name.  This can lead to a lot of unnecessarily heavy thinking
+when following some very typical stack traces.  It also means that the name of
+the resource being declared is often not on the stack, making debugging harder.
+
+Here we propose that all resource and definition DSL be added to
+`Chef::DSL::Recipe` as actual methods, as they are created, so that they can be
+subject to inspection and the stack traces can be followed.  This will happen
+automatically when the user declares resources and definitions.
+
+### Automatically add new resources and providers to DSL, regardless of namespace
+
+Presently, when you type `my_resource` in the Chef DSL, in the absence of any
+`provides` directives to direct Chef to the correct resource class, the
+namespaces `Chef::Resource` is explicitly searched for
+`Chef::Resource::MyResource`.
+
+This means that if you have seen examples of classes in Chef::Resource, and want
+to do the obvious thing and just move to your own namespace, it doesn't
+immediately work and it is far from apparent why. The result? A massive traffic
+jam of non-core resources in the `Chef::Resource` namespace and puzzling error
+messages for cookbook conflicts such as "superclass mismatch for class MyResource."
+
+Chef will now automatically call `provides` on all new, named resource classes,
```

Comment:
I actually favor dropping the implicit magical class based stuff entirely and requiring at least one 'provides'.  For LWRPs and other DSLey ways of writing resources this can be done for the ultimate end user, but HWRP users should need to specify a provides line.  Obviously, this is a breaking change, Chef-13, yadda, yadda.

For back-compat we could wire up a define_method on inheriting from Chef::Resource if the resource is under the Chef::Resource namespace.  That method should then spit out a deprecation warning if its run.  Then a subsequent explicit provides line would define_method and overwrite the magically added one and would not spit out a deprecation warning.  Eventually the magic on inheriting from Chef::Resource gets removed.